packages/x-tree-view-pro/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.ts

```diff
@@ -0,0 +1,296 @@
+        const childrenFetchError = error as Error;
+        // handle errors here
+        instance.setDataSourceError(id, childrenFetchError);
+        instance.removeChildren(id);
+      } finally {
+        // unset loading
+        instance.setDataSourceLoading(id, false);
+
+        nestedDataManager.setRequestSettled(id);
+      }
+    },
+    [nestedDataManager, params.dataSource, instance, cache, store],
+  );
+
+  useInstanceEventHandler(instance, 'beforeItemToggleExpansion', async (eventParameters) => {
+    if (!isLazyLoadingEnabled) {
+      return;
+    }
+
+    if (selectorIsItemExpanded(store.value, eventParameters.itemId)) {
+      return;
+    }
+
+    eventParameters.isExpansionPrevented = true;
+    await instance.fetchItems([eventParameters.itemId]);
+    const fetchErrors = Boolean(selectorGetTreeItemError(store.value, eventParameters.itemId));
+    if (!fetchErrors) {
+      instance.setItemExpansion(eventParameters.event, eventParameters.itemId, true);
+      if (selectorIsItemSelected(store.value, eventParameters.itemId)) {
+        // make sure selection propagation works correctly
+        instance.selectItem({
+          event: eventParameters.event as React.SyntheticEvent,
+          itemId: eventParameters.itemId,
+          keepExistingSelection: true,
+          shouldBeSelected: true,
+        });
+      }
+    }
+  });
+
+  React.useEffect(() => {
+    if (isLazyLoadingEnabled && firstRenderRef.current) {
+      if (params.items.length) {
+        const getChildrenCount = params.dataSource?.getChildrenCount || (() => 0);
+        instance.addItems({ items: params.items, depth: 0, getChildrenCount });
+      } else {
+        instance.fetchItems();
+      }
+      firstRenderRef.current = false;
+    }
+  }, [instance, params.items, params.dataSource, isLazyLoadingEnabled]);
+
+  const pluginContextValue = React.useMemo(
+    () => ({ lazyLoading: params.dataSource !== undefined }),
+    [params.dataSource],
+  );
+
+  if (!isLazyLoadingEnabled) {
+    return {} as UseTreeViewLazyLoadingSignature;
+  }
+  instance.preventItemUpdates();
+
+  return {
+    instance: {
+      fetchItemChildren,
+      fetchItems,
+      setDataSourceLoading,
+      setDataSourceError,
+    },
+    publicAPI: {},
+    contextValue: pluginContextValue,
+  };
+};
+
+useTreeViewLazyLoading.getDefaultizedParams = ({ params, experimentalFeatures }) => {
+  const canUseFeature = experimentalFeatures?.lazyLoading;
+  if (process.env.NODE_ENV !== 'production') {
+    if (params.dataSource && !canUseFeature) {
+      warnOnce([
+        'MUI X: The label editing feature requires the `lazyLoading` experimental feature to be enabled.',
+        'You can do it by passing `experimentalFeatures={{ lazyLoading: true}}` to the Rich Tree View Pro component.',
+        'Check the documentation for more details: https://mui.com/x/react-tree-view/rich-tree-view/lazy-loading/',
+      ]);
+    }
+  }
+  const defaultDataSource = params?.dataSource || {
```

Comment:
Nitpick, I think it's a good practice to always use `??`, that way the day your value can be `0` or `''` you don't even have to think about it

```suggestion
  const defaultDataSource = params?.dataSource ?? {
```